ssl/quic/quic_cfq.c

```diff
@@ -0,0 +1,345 @@
+    void                   *free_cb_arg;
+    uint64_t                frame_type;
+    size_t                  encoded_len;
+    uint32_t                priority, pn_space;
+    char                    state;
+};
+
+uint64_t ossl_quic_cfq_item_get_frame_type(QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    return ex->frame_type;
+}
+
+const unsigned char *ossl_quic_cfq_item_get_encoded(QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    return ex->encoded;
+}
+
+size_t ossl_quic_cfq_item_get_encoded_len(QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    return ex->encoded_len;
+}
+
+int ossl_quic_cfq_item_get_state(QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    return ex->state;
+}
+
+uint32_t ossl_quic_cfq_item_get_pn_space(QUIC_CFQ_ITEM *item)
+{
+    QUIC_CFQ_ITEM_EX *ex = (QUIC_CFQ_ITEM_EX *)item;
+
+    return ex->pn_space;
+}
+
+typedef struct quic_cfq_item_list_st {
+    QUIC_CFQ_ITEM_EX *head, *tail;
+} QUIC_CFQ_ITEM_LIST;
+
+struct quic_cfq_st {
+    /* 
+     * Invariant: A CFQ item is always in exactly one of these lists, never more
+     * or less than one.
+     *
+     * Invariant: The list the CFQ item is determined exactly by the state field
+     * of the item.
+     */
+    QUIC_CFQ_ITEM_LIST                      new_list, tx_list, free_list;
+};
+
+static int compare(const QUIC_CFQ_ITEM_EX *a, const QUIC_CFQ_ITEM_EX *b)
+{
+    if (a->pn_space < b->pn_space)
+        return -1;
+    else if (a->pn_space > b->pn_space)
+        return 1;
+
+    if (a->priority < b->priority)
+        return -1;
+    else if (a->priority > b->priority)
+        return 1;
+
+    return 0;
+}
+
+static void list_remove(QUIC_CFQ_ITEM_LIST *l, QUIC_CFQ_ITEM_EX *n)
+{
+    if (l->head == n)
+        l->head = n->next;
+    if (l->tail == n)
+        l->tail = n->prev;
+    if (n->prev != NULL)
+        n->prev->next = n->next;
+    if (n->next != NULL)
+        n->next->prev = n->prev;
+    n->prev = n->next = NULL;
+}
+
+static void list_insert_head(QUIC_CFQ_ITEM_LIST *l, QUIC_CFQ_ITEM_EX *n)
+{
+    n->next = l->head;
+    n->prev = NULL;
+    l->head = n;
+    if (n->next != NULL)
+        n->next->prev = n;
+    if (l->tail == NULL)
+        l->tail = n;
+}
+
+static void list_insert_tail(QUIC_CFQ_ITEM_LIST *l, QUIC_CFQ_ITEM_EX *n)
+{
+    n->prev = l->tail;
+    n->next = NULL;
+    l->tail = n;
+    if (n->prev != NULL)
+        n->prev->next = n;
+    if (l->head == NULL)
+        l->head = n;
+}
+
+static void list_insert_after(QUIC_CFQ_ITEM_LIST *l,
+                              QUIC_CFQ_ITEM_EX *ref,
+                              QUIC_CFQ_ITEM_EX *n)
+{
+    n->prev = ref;
+    n->next = ref->next;
+    if (ref->next != NULL)
+        ref->next->prev = n;
+    ref->next = n;
+    if (l->tail == ref)
+        l->tail = n;
+}
+
+static void list_insert_sorted(QUIC_CFQ_ITEM_LIST *l, QUIC_CFQ_ITEM_EX *n,
```

Comment:
Could the worst case O(n²) behaviour be triggered by an attacker?

I know the use case here avoids the pathological case for normal operation but attackers aren't known for following normal operation.

Reply:
The attacker would have to have a way to force us to generate a large number of NEW_CONN_ID, RETIRE_CONN_ID or NEW_TOKEN frames. There is no way for a peer to solicit generation of a NEW_TOKEN frame, and there are mechanisms for limiting the number of connection IDs in play, so this doesn't seem too concerning.